File: api/src/test/java/net/blueshell/api/testsupport/GenericsPredicates.java
package net.blueshell.api.testsupport;

import com.tngtech.archunit.base.DescribedPredicate;
import com.tngtech.archunit.core.domain.JavaClass;
import com.tngtech.archunit.core.domain.JavaParameterizedType;
import com.tngtech.archunit.core.domain.JavaType;
import com.tngtech.archunit.core.domain.JavaWildcardType;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Helpers for ArchUnit predicates involving generic return types.
 */
public final class GenericsPredicates {

    private GenericsPredicates() {
    }

    public static DescribedPredicate<JavaType> assignableToGeneric(Class<?> raw, Class<?>... typeArgs) {
        return assignableToGeneric(raw, false, typeArgs);
    }

    public static DescribedPredicate<JavaType> assignableToGeneric(Class<?> raw, boolean allowRaw, Class<?>... typeArgs) {
        String description = describe(raw, typeArgs);
        return DescribedPredicate.describe("assignable to %s".formatted(description), type -> {
            JavaClass erasure = type.toErasure();
            if (!erasure.isAssignableTo(raw)) return false;

            if (!(type instanceof JavaParameterizedType parameterized)) {
                return allowRaw;
            }

            List<JavaType> actualArgs = parameterized.getActualTypeArguments();
            if (actualArgs.size() != typeArgs.length) return false;

            for (int i = 0; i < typeArgs.length; i++) {
                JavaType actual = actualArgs.get(i);
                Class<?> expected = typeArgs[i];
                if (!typeArgMatches(actual, expected)) return false;
            }
            return true;
        });
    }

    private static boolean typeArgMatches(JavaType actual, Class<?> expected) {
        if (actual instanceof JavaWildcardType wildcard) {
            var uppers = wildcard.getUpperBounds();
            if (uppers.isEmpty()) {
                return expected == Object.class;
            }
            return uppers.stream().anyMatch(u -> u.toErasure().isAssignableTo(expected));
        }
        return actual.toErasure().isAssignableTo(expected);
    }

    private static String describe(Class<?> raw, Class<?>... typeArgs) {
        if (typeArgs == null || typeArgs.length == 0) return raw.getSimpleName();
        String args = Arrays.stream(typeArgs).map(Class::getSimpleName).collect(Collectors.joining(", "));
        return "%s<%s>".formatted(raw.getSimpleName(), args);
    }
}

File: api/src/test/kotlin/net/blueshell/api/testsupport/GenericsPredicates.kt
package net.blueshell.api.testsupport

import com.tngtech.archunit.base.DescribedPredicate
import com.tngtech.archunit.core.domain.JavaClass
import com.tngtech.archunit.core.domain.JavaParameterizedType
import com.tngtech.archunit.core.domain.JavaType
import com.tngtech.archunit.core.domain.JavaWildcardType

/**
 * Helpers for ArchUnit predicates involving generic return types.
 */
object GenericsPredicates {

    fun assignableToGeneric(raw: Class<*>, vararg typeArgs: Class<*>): DescribedPredicate<JavaType> =
        assignableToGeneric(raw = raw, allowRaw = false, typeArgs = typeArgs)

    fun assignableToGeneric(
        raw: Class<*>,
        allowRaw: Boolean,
        vararg typeArgs: Class<*>
    ): DescribedPredicate<JavaType> {
        val description = describe(raw, typeArgs.toList())
        return DescribedPredicate.describe("assignable to $description") { type ->
            val erasure: JavaClass = type.toErasure()
            if (!erasure.isAssignableTo(raw)) return@describe false

            val parameterized = type as? JavaParameterizedType ?: return@describe allowRaw
            val actualArgs = parameterized.actualTypeArguments
            if (actualArgs.size != typeArgs.size) return@describe false

            actualArgs.zip(typeArgs.asList()).all { (actual, expected) ->
                typeArgMatches(actual, expected)
            }
        }
    }

    private fun typeArgMatches(actual: JavaType, expected: Class<*>): Boolean {
        val wildcard = actual as? JavaWildcardType
        if (wildcard != null) {
            val uppers = wildcard.upperBounds
            if (uppers.isEmpty()) return expected == Any::class.java
            return uppers.any { it.toErasure().isAssignableTo(expected) }
        }
        return actual.toErasure().isAssignableTo(expected)
    }

    private fun describe(raw: Class<*>, typeArgs: List<Class<*>>): String {
        if (typeArgs.isEmpty()) return raw.simpleName
        val args = typeArgs.joinToString(", ") { it.simpleName }
        return "${raw.simpleName}<$args>"
    }
}
